src/nf/nf_optimizers.f90

```diff
@@ -123,4 +150,45 @@ pure subroutine minimize_rmsprop(self, param, gradient)
 
   end subroutine minimize_rmsprop
 
+
+  impure elemental subroutine init_adam(self, num_params)
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
+      allocate(self % m(num_params), self % v(num_params))
+      self % m = 0
+      self % v = 0
+    end if
+  end subroutine init_adam
+
+
+  pure subroutine minimize_adam(self, param, gradient)
+    !! Concrete implementation of an Adam optimizer update rule.
+    class(adam), intent(inout) :: self
+    real, intent(inout) :: param(:)
+    real, intent(in) :: gradient(:)
+
+    self % t = self % t + 1
+
+    ! If weight_decay > 0, use L2 regularization;
+    ! otherwise, default to regular Adam.
+    associate(g => gradient + 2 * self % weight_decay * param)
```

Comment:
Why do you have a factor of 2 here? In the [paper, Algorithm 2, line 6](https://arxiv.org/pdf/1711.05101.pdf) I see just $\lambda \theta_{t-1}$ (i.e. `self % weight_decay * param`).

Another question is, based on my understanding from the paper, applying `weight_decay * param` here is L2 regularization. However, for decoupled weight decay regularization, it seems like it should be added below on L188.

Reply:
You're right, the scaling factor of 2 is not present in the paper. The reason was when the weight decay coefficient is large, the term `weight_decay * param` can become very large.

Please review the recent changes.

Reply:
Thanks @Spnetic-5. We still need to decide how to handle the two different regularization approaches (L2 and decoupled). We currently have both enabled if `weight_decay > 0`. However, I think only one or the other should be applied at a time. PyTorch and Keras differentiate the two by providing two separate classes: Adam and AdamW. I think this is unnecessary in our case. We can have just Adam for now, and allow two different regularization approaches to be applied via input arguments. What do you think about using two separate weight decays to control this, i.e.:

```fortran
  real :: weight_decay_l2 = 0  ! L2 regularization (Adam)
  real :: weight_decay_decoupled = 0 ! decoupled weight decay regularization (AdamW)
```

Reply:
Yup, I agree that having two separate input `weight_decay` arguments isa good approach. So by default, both weight decays will be set to 0, so no regularization will be applied. I'll go ahead and do the changes.